### PR TITLE
Wrong spectrogram scaling in inference script

### LINKED ISSUE #3: Much worse results than in paper
Hi,

First of all, thank you for this nice implementation. 

I trained the network with default settings and data (~500k iteration), but the results are really unnaturalistic (eg.: [link](https://drive.google.com/file/d/1CFkTlAbdV6qWWaK6xmB7qRzido1tKFr2/view?usp=sharing )) and far from the samples provided by the author of the paper:
http://www.kecl.ntt.co.jp/people/kaneko.takuhiro/projects/maskcyclegan-vc/index.html

Why is this? Did you experience the same or you got nice results?

### PR DESCRIPTION
**Pull request by terbed**
In the inference script the scaling back with speaker statistics is wrong:
```
            wav_fake_B = decode_melspectrogram(self.vocoder, fake_B[0].detach(
            ).cpu(), self.dataset_A_mean, self.dataset_A_std).cpu()
```
This should be scaled with B speaker statistics. 

This correction largely improves the performance of the inference.

Solves issue #3 .

### CODE REVIEW

File: args/train_arg_parser.py [hunk 1]
```diff
@@ -16,11 +16,11 @@ def __init__(self):
         super(TrainArgParser, self).__init__()
         self.isTrain = True
 
-        self.parser.add_argument('--num_epochs', type=int, default=20, help='Number of epochs to train.')
+        self.parser.add_argument('--num_epochs', type=int, default=6500, help='Number of epochs to train.')
         self.parser.add_argument(
-            '--lr', type=float, default=5e-4, help='Learning rate.')
+            '--decay_after', type=float, default=2e5, help='Learning rate.')
```
Comment (hikaruhotta) [inline args/train_arg_parser.py]: Please change this comment to "Decay learning rate after n iterations."

File: args/train_arg_parser.py [hunk 2]
```diff
@@ -16,11 +16,11 @@ def __init__(self):
...
         self.parser.add_argument(
-            '--decay_after', type=float, default=1e4, help='Learning rate.')
+            '--stop_identity_after', type=float, default=1e4, help='Learning rate.')
```
Comment (hikaruhotta) [inline args/train_arg_parser.py]: Please change this comment to "Stop using identity loss after n iterations,"

File: data_preprocessing/preprocess.py [hunk 3]
```diff
@@ -31,8 +31,10 @@ def normalize_mel(wavspath):
     mel_list = list()
     for wavpath in tqdm(wav_files, desc='Preprocess wav to mel'):
         wav_orig, _ = librosa.load(wavpath, sr=SAMPLING_RATE, mono=True)
+        # wav_orig, _ = librosa.effects.trim(wav_orig, top_db=15)  # clip silent both sides
         spec = vocoder(torch.tensor([wav_orig]))
-        mel_list.append(spec.cpu().detach().numpy()[0])
+        if spec.shape[-1] >= 64:
```
Comment (hikaruhotta) [inline data_preprocessing/preprocess.py]: Add comment:

// training sample consists of 64 randomly cropped frames

File: data_preprocessing/preprocess.py [hunk 4]
```diff
@@ -31,8 +31,10 @@ def normalize_mel(wavspath):
     mel_list = list()
     for wavpath in tqdm(wav_files, desc='Preprocess wav to mel'):
         wav_orig, _ = librosa.load(wavpath, sr=SAMPLING_RATE, mono=True)
+        # wav_orig, _ = librosa.effects.trim(wav_orig, top_db=15)  # clip silent both sides
```
Comment (hikaruhotta) [inline data_preprocessing/preprocess.py]: Please remove commented out code

File: mask_cyclegan_vc/model.py [hunk 5]
```diff
@@ -327,11 +327,11 @@ def __init__(self, input_shape=(80, 64), residual_in_channels=256):
                                                        padding=[0, 1]))
 
     def downsample(self, in_channels, out_channels, kernel_size, stride, padding):
-        convLayer = nn.Sequential(nn.Conv2d(in_channels=in_channels,
+        convLayer = nn.Sequential(nn.utils.spectral_norm(nn.Conv2d(in_channels=in_channels,
```
Comment (hikaruhotta) [inline mask_cyclegan_vc/model.py]: Please remove spectral norm for now. Perhaps it can be added as an arg?

File: mask_cyclegan_vc/test.py [hunk 6]
```diff
@@ -82,17 +83,40 @@ def loadPickleFile(self, fileName):
...
+        for i, sample in enumerate(tqdm(self.test_dataloader)):
+
             save_path = None
             if self.model_name == 'generator_A2B':
-                save_path = os.path.join(self.converted_audio_dir, f"converted_{self.speaker_A_id}_to_{self.speaker_B_id}{i}.wav")
+                real_A = sample
+                real_A = real_A.to(self.device, dtype=torch.float)
+                fake_B = self.generator(real_A, torch.ones_like(real_A))
+
+                wav_fake_B = decode_melspectrogram(self.vocoder, fake_B[0].detach(
```
Comment (hikaruhotta) [inline mask_cyclegan_vc/test.py:94]: great great great catch!

File: mask_cyclegan_vc/train.py [hunk 7]
```diff
@@ -287,53 +305,57 @@ def train(self):
...
+                    # Log Mel-spectrograms .png
+                    real_mel_A_fig = get_mel_spectrogram_fig(
+                        real_A[0].detach().cpu())
+                    fake_mel_A_fig = get_mel_spectrogram_fig(
+                        generated_A[0].detach().cpu())
+                    real_mel_B_fig = get_mel_spectrogram_fig(
+                        real_B[0].detach().cpu())
+                    fake_mel_B_fig = get_mel_spectrogram_fig(
+                        generated_B[0].detach().cpu())
+                    self.logger.visualize_outputs({"real_voc_spec": real_mel_A_fig, "fake_coraal_spec": fake_mel_B_fig,
```
Comment (hikaruhotta) [inline mask_cyclegan_vc/train.py]: Names should be "fake_A_spec" and "fake_B_spec" <- assigned @HikaruHotta

**Review by hikaruhotta — CHANGES_REQUESTED** [review]
Thank you for finding that bug in the testing script. That explains the discrepancies I was seeing from getting the audio samples directly off of tensorboard and those from the test script. I noticed that the file bash_scripts/mask_cyclegan_test.sh was deleted. Was there a reason for this? Otherwise, I will add it back. I've made a couple of minor comments which can be resolved very easily. I appreciate how you touched up the training script with torch.no_grad and .eval() .train().

**Comment by terbed** [comment]
> I noticed that the file bash_scripts/mask_cyclegan_test.sh was deleted. Was there a reason for this? Otherwise, I will add it back. 

No, it was accidental, thanks!

Thank you for the constructive comments, I will include them!

File: mask_cyclegan_vc/model.py (hunk 5 above)
Comment (terbed) [reply mask_cyclegan_vc/model.py]: This modification not resulted in perceptible improvement, therefore I remove it.

**Review by hikaruhotta — CHANGES_REQUESTED** [review]
Juts one minor change. I'm currently running training on your version so I will approve these changes after training concludes

**Review by hikaruhotta — APPROVED** [review]
LGTM!
